fetch/fetch.go

```diff
@@ -0,0 +1,526 @@
+	ID        types.Hash32
+	Responses []responseMessage
+}
+
+// Config is the configuration file of the Fetch component
+type Config struct {
+	BatchTimeout      int
+	MaxRetiresForPeer int
+	BatchSize         int
+	RequestTimeout    int
+}
+
+type peers interface {
+	GetPeers() []p2ppeers.Peer
+}
+
+// MessageNetwork is a network interface that allows fetch to communicate with other nodes with 'fetch servers'
+type MessageNetwork struct {
+	*server.MessageServer
+	peers
+	log.Log
+}
+
+// NewMessageNetwork creates a new instance of the fetch network server
+func NewMessageNetwork(requestTimeOut int, net service.Service, protocol string, log log.Log) *MessageNetwork {
+	return &MessageNetwork{
+		server.NewMsgServer(net.(server.Service), protocol, time.Duration(requestTimeOut)*time.Second, make(chan service.DirectMessage, p2pconf.Values.BufferSize), log),
+		p2ppeers.NewPeers(net, log.WithName("peers")),
+		log,
+	}
+}
+
+// GetRandomPeer returns a random peer from current peer list
+func (f MessageNetwork) GetRandomPeer() p2ppeers.Peer {
+	peers := f.peers.GetPeers()
+	if len(peers) == 0 {
+		f.Log.Panic("cannot send fetch - no peers found")
+	}
+	rand.Seed(time.Now().Unix()) // initialize global pseudo random generator
+	return peers[rand.Intn(len(peers))]
+}
+
+type network interface {
+	GetRandomPeer() p2ppeers.Peer
+	SendRequest(msgType server.MessageType, payload []byte, address p2pcrypto.PublicKey, resHandler func(msg []byte), failHandler func(err error)) error
+}
+
+// Fetch is the main struct that contains network peers and logic to batch and dispatch hash fetch Requests
+type Fetch struct {
+	cfg                  Config
+	log                  log.Log
+	dbs                  map[Hint]database.Database
+	activeRequests       map[types.Hash32][]request
+	activeBatches        map[types.Hash32]requestBatch
+	net                  network
+	requestReceiver      chan request
+	batchRequestReceiver chan []request
+	batchTimeout         *time.Ticker
+	stop                 chan struct{}
+	activeReqM           sync.RWMutex
+	activeBatchM         sync.RWMutex
+}
+
+// NewFetch creates a new FEtch struct
+func NewFetch(cfg Config, network service.Service, logger log.Log) *Fetch {
+
+	srv := NewMessageNetwork(cfg.RequestTimeout, network, fetchProtocol, logger)
+
+	f := &Fetch{
+		cfg:             cfg,
+		log:             logger,
+		dbs:             make(map[Hint]database.Database),
+		activeRequests:  make(map[types.Hash32][]request),
+		net:             srv,
+		requestReceiver: make(chan request),
+		batchTimeout:    time.NewTicker(time.Second * time.Duration(cfg.BatchTimeout)),
+		stop:            make(chan struct{}),
+		activeBatches:   make(map[types.Hash32]requestBatch),
+	}
+
+	srv.RegisterBytesMsgHandler(fetch, f.FetchRequestHandler)
+
+	return f
+}
+
+// Start handling fetch Requests
+func (f *Fetch) Start() {
+	go f.loop()
+}
+
+// Stop handling fetch Requests
+func (f *Fetch) Stop() {
+	f.batchTimeout.Stop()
+	f.stop <- struct{}{}
```

Comment:
You added an additional boolean `stopped` flag, but it's not indicating the stop completion, but rather stop request, similarly to the chan. Instead of having 2 sources of truth, you can just have a boolean abstraction for the chan.
```go
func stopped() bool {
  select {
    case <-f.stopped
      return true 
    default:
      return false
  }
}
```
Assuming channel will be closed.